### PR TITLE
egui_extras: enable virtual scroll for heterogenous rows

### PR DESCRIPTION
**Pull request by waynr**
Introduce `TableBody.heterogenous_rows` and `TableRowBuilder`, the former of which takes as an argument the latter. `TableRowBuilder` provides two methods that enable virtual scrolling for rows with non-uniform heights. Those methods are:

* `TableRowBuilder.row_heights` which returns an iterator over `f32` to allow incremental virtual scroll buffer calculation.
* `TableRowBuilder.populate_row` which `TableBody.heterogenous_rows` uses to allow `TableRowBuilder` implementations to, you guessed it, populate rows that are visible.

One thought that occurs to me while writing this description is that `TableBody.heterogenous_rows` could look more like the following:

```
    pub fn heterogenous_rows(
        mut self,
        row_heights: impl Iterator<Item = f32> + '_,
        mut row: impl FnMut(usize, TableRow<'_, '_>),
    )
```

This could potentially be easier to use, considering all the trouble I had coming up with and implementing the trait. Happy to make this change if the maintainers prefer.

Closes <https://github.com/emilk/egui/issues/THE_RELEVANT_ISSUE>.

### CODE REVIEW

File: egui_demo_lib/src/apps/demo/table_demo.rs
Comment (waynr) [inline egui_demo_lib/src/apps/demo/table_demo.rs:79]: note: technically both branches here use virtual scroll, so it might be helpful to come up with a different distinction here. while doing that it could also be helpful from a development standpoint to make the number of rows (in either case) configurable to make it easier to test edge cases

File: egui_extras/src/table.rs
Comment (waynr) [inline egui_extras/src/table.rs:355]: note for reviewers: widths isn't used in the demo app, but I included it here because my main use case for the `Table` type is to implement a chat box where text will be wrapped based on width and so height will be tied to cell width.

actually, come to think of it this was the main reason i thought it would be good to use a trait rather than simply taking an iterator over `f32` without any input from `heterogenous_rows`.

Comment (waynr) [reply egui_extras/src/table.rs:355]: if there were a `widths(&self) -> Vec<32>` method on `TableBody` then coming up with an iterator of floating point height values wouldn't explicitly require it as an input

File: egui_extras/src/table.rs
Comment (waynr) [inline egui_extras/src/table.rs:375]: note: just occurred to me that breaking this into three separate loops might make it a bit easier to reason about. can use breaks at the boundaries between non-visible and visible areas.

**Comment by waynr** [comment]
For what it's worth, when I ran this in the demo app using 100_000 rows in the heterogeneous rows (same as in the homogeneous rows example) I noticed that CPU time / frame seemed to average around 9 or 10 ms on my computer vs the 4 or 5 of the homogeneous rows example. Of course this is many orders of magnitude better than the heterogeneous example at 100_000 rows prior to this PR...

File: egui_extras/src/table.rs
Comment (waynr) [inline egui_extras/src/table.rs:385]: TODO: fix out of date comment

File: egui_demo_lib/src/apps/demo/table_demo.rs
Comment (emilk) [reply egui_demo_lib/src/apps/demo/table_demo.rs:79]: I think it would be good to demo all three version - the simple version for short tables, the version with homogeneous row heights, and the new heterogeneous version

File: egui_extras/src/table.rs
Comment (emilk) [inline egui_extras/src/table.rs:380]: Accumulating `f32` will lead to precision issues really quickly (even worse than those [already present](https://github.com/emilk/egui/issues/1391)). Better to use `f64` for `height_above_visible`,  `height_below_visible` and `current_height`.

File: egui_extras/src/table.rs
Comment (emilk) [inline egui_extras/src/table.rs:461]: needs a docstring or a better name (`advance_height`?)

File: egui_extras/src/table.rs
Comment (emilk) [inline egui_extras/src/table.rs:360]: Needs a docstring with an example

File: egui_extras/src/table.rs
Comment (emilk) [reply egui_extras/src/table.rs:355]: I agree that using a trait for this is quite cumbersome. It would be much nicer to use with just closures.

Adding `fn width(&self) -> &[f32]` to `TableBody` is easy enough.

**Review by emilk — CHANGES_REQUESTED** [review]
This is a really good addition! But let's see if we can do it without the `trait`!

File: egui_extras/src/table.rs
Comment (waynr) [reply egui_extras/src/table.rs]: Should the `height` values be given in `f64` then also? I don't know much about precision issues between different bit widths. But I did go ahead and make this change.

File: egui_extras/src/table.rs
Comment (waynr) [reply egui_extras/src/table.rs]: I added a docstring and removed the `pub`, but `buffer` still makes more sense to me semantically -- but since you're the maintainer if you have a strong preference I'll go with whatever your suggestion is.

File: egui_extras/src/table.rs
Comment (emilk) [reply egui_extras/src/table.rs]: No, `height` can be `f32`. It is the accumulator that needs to be wide.

File: egui_extras/src/table.rs
Comment (emilk) [reply egui_extras/src/table.rs]: buffer is so many different things (including a verb). `add_buffer` or `add_space`?

File: egui_demo_lib/src/apps/demo/table_demo.rs
Comment (emilk) [inline egui_demo_lib/src/apps/demo/table_demo.rs:35]: You should be able to just put the settings before the `StripBuilder`. We only need `StripBuilder` here because of the table uses `remainder`.

File: egui_demo_lib/src/apps/demo/table_demo.rs
Comment (emilk) [inline egui_demo_lib/src/apps/demo/table_demo.rs:46]: These are 2*2=4 options for three different demos. Perhaps use an enum and radio buttons, or only show `heterogeneous_rows` in case `virtual_scroll==true`

File: egui_extras/src/table.rs
Comment (emilk) [inline egui_extras/src/table.rs:357]: I think this is a weird name and a weird behavior - the _negative_ amount of y progress?

I think this would be better as

```suggestion
    fn y_progress(&self) -> f32 {
        self.layout.current_y() - self.start_y
    }
```

so that the returned value is positive (or am I misunderstanding it?)

File: egui_extras/src/table.rs
Comment (emilk) [inline egui_extras/src/table.rs:440]: I don't quite follow the logic of everything, but it seems like this line should be taking `height` into account. If the new row height is so wide that it should be visible, then we should `break` and paint it. I think if you do that, then we can remove `VIRTUAL_EXTENSION` completely.

File: egui_extras/src/table.rs
Comment (emilk) [inline egui_extras/src/table.rs:476]:
```suggestion
        // if height below visible is > 0 here then we need to add a buffer to allow the table to
```

**Review by emilk — CHANGES_REQUESTED** [review]

File: egui_demo_lib/src/apps/demo/table_demo.rs
Comment (waynr) [reply egui_demo_lib/src/apps/demo/table_demo.rs:35]: I don't know any other way to get a horizontal layout (checkbox options on the left, slider on the right). If I move the setting strip builder with the horizontal layout outside of the top-level strip builder than the whole table demo window automatically expands in height well below the bottom edge of the application window.

Anyway, if this is a demo what's the problem with demonstrating how to embed a horizontally-oriented strip inside the cell of a vertically-oriented strip?

File: egui_extras/src/table.rs
Comment (parasyte) [inline egui_extras/src/table.rs:364]: This can return a slice.

```suggestion
    pub fn widths(&self) -> &[f32] {
```

Comment (waynr) [reply egui_extras/src/table.rs:364]: I'm still fairly new to rust so I don't fully understand the signficance -- is there a reason to prefer one over the other?

File: egui_extras/src/table.rs
Comment (waynr) [reply egui_extras/src/table.rs:357]: I could also be misunderstanding, but if you look at the way [delta was calculated prior to this PRd in `TableBody::rows`](https://github.com/emilk/egui/blob/5dff1e42c6ec8a3740c9cf0b18400743d73a0266/egui_extras/src/table.rs#L359-L363) the result was expected to be negative if any buffer spaces was needed above the visible range. My intent with this `delta` helper method was to clean that up a little and to reuse the calculation in `TableBody::heterogeneous_rows`.

So I agree with the intent of your feedback here (ie that it's easier to reason about a positive result), just not with the suggested change since to my understanding what I've written should always result in >= 0 value for delta. But my understanding is definitely limited since I don't have a good conceptual understanding of the difference between `self.start_y` and `self.layout.current_y()` nor even where the origin of the coordinate system in egui is. My guess is that `self.start_y` is the coordinate of the top of the visible rectangle and `self.layout.current_y()` is the top of the "virtual" rectangle pushed upward by the `TableRow`. Or something like that. All I know is that if I flip the subtraction operands around then I have to either negate the result or flip a bunch of comparison operands.

I'll definitely make the suggested name change though.

File: egui_extras/src/table.rs
Comment (waynr) [reply egui_extras/src/table.rs:440]: While I think this suggestion is ultimately correct -- it solve the problem in the demo app where row index 0 isn't considered an "extra thick row" despite `0 % 6 == 0` (the logic that determines which rows should be thicker) -- it doesn't solve the same problem that I intended to solve with `VIRTUAL_EXTENSION`, which is that when scrolling up rows abruptly disappear rather than slide "under" the lip of the visible rectangle area. In theory even this `VIRTUAL_EXTENSION` constant may not work for sufficiently tall rows...

File: egui_demo_lib/src/apps/demo/table_demo.rs
Comment (waynr) [reply egui_demo_lib/src/apps/demo/table_demo.rs]: I took the approach of only showing the `heterogeneous_rows` setting if `virtual_scroll==true`.

File: egui_extras/src/table.rs
Comment (parasyte) [reply egui_extras/src/table.rs]: There a few reasons to return a slice. One is that it provides an abstraction that hides implementation details. For instance the list of widths could internally be stored in any collection which can Deref to slice and the type could be changed freely without changing the API.

Also if you look through the docs for `Vec` you'll see that it only implements very few methods that take `&self`, and the useful ones are also implemented on slice.

Lastly, it's just more idiomatic. In the same way you would pass a slice as an argument instead of a borrowed vector. A slice works with multiple types of collections.

Comment (waynr) [reply egui_extras/src/table.rs]: That's super helpful to know, thanks for the explanation! I did make the suggested change fyi.

Comment (parasyte) [reply egui_extras/src/table.rs]: Glad to help!

File: egui_extras/src/table.rs
Comment (emilk) [inline egui_extras/src/table.rs:438]: I believe `VIRTUAL_EXTENSION` could be removed if we changed this logic to take `height` into account (`if height_above_visible + height as f64 >= …`).

(PS: I thought I already made this comment, but now I can't find it on GitHub…)

**Comment by emilk** [comment]
You don't need to wait for CI - just run `./sh/check.sh` (https://github.com/emilk/egui/blob/master/CONTRIBUTING.md#making-a-pr)

**Comment by waynr** [comment]
@emilk
```
<wayne@thing3:~/projects/egui>
zsh/3 4901 [101]  (git)-[enable-virtual-scroll-for-heterogeneous-rows]-% ./sh/check.sh
./sh/check.sh: 2: Bad substitution
+ cargo check --workspace --all-targets
error: could not find `Cargo.toml` in `/home/wayne/projects` or any parent directory
```

**Comment by emilk** [comment]
> @emilk
> 
> ```
> <wayne@thing3:~/projects/egui>
> zsh/3 4901 [101]  (git)-[enable-virtual-scroll-for-heterogeneous-rows]-% ./sh/check.sh
> ./sh/check.sh: 2: Bad substitution
> + cargo check --workspace --all-targets
> error: could not find `Cargo.toml` in `/home/wayne/projects` or any parent directory
> ```

Oh, I guess bash scripts aren't very portable 😭

**Comment by waynr** [comment]
> Oh, I guess bash scripts aren't very portable 

Technically these scripts aren't using bash unless the user has `sh` linked to `/bin/bash`: https://github.com/emilk/egui/blob/68d5806b41725fb8f8a5e62d7acd94b5ac9140f0/sh/check.sh#L1

**Review by emilk — APPROVED** [review]

**Comment by emilk** [comment]
Thanks!

**Pull request merged**
